metagamer/environments/qagent.py

```diff
@@ -1,144 +1,432 @@
+        """
+        # choice to prune or build
+        # we choose prune here since it's less likely to go wrong
+        super().__init__()
+        self.qdict: Dict[str, Dict[int, float]] = collections.defaultdict(
+            lambda: {i: 0 for i in range(9)}
+        )
+        # update decay - smaller alphas have longer memory, larger alphas shorter
+        self.alpha = alpha
+        # discount factor when updating reward - high gamma is more greedy, low gamma is more patient
+        self.discount = gamma
 
-    def train(self, num_epsiodes, initeps=1, finaleps=0.05):
+    def set_reward(self, state: str, action: int, value: float):
         """
-        Simple linear drop for epsilon.
+        Only allowed to set a state, value pair.
+        Update is as per alpha value
         """
-        epsdecay = (initeps - finaleps) / num_epsiodes
-        epsilon = initeps
+        if self.eval:
+            # no learning in eval mode
+            return
+        self.qdict[state][action] = (
+            self.qdict[state][action] * (1 - self.alpha) + value * self.alpha
+        )
+
+    def prune_qdict(self, state: str, valid_actions: Sequence[int]):
+        """Optional function that allows us to reduce space"""
+        if set(valid_actions) - self.qdict[state].keys():
+            raise ValueError("We have found valid actions not in our qtable for state")
+        elif self.qdict[state].keys() - set(valid_actions):
+            # we can prune down since we don't need the other moves
+            self.qdict[state] = {k: self.qdict[state][k] for k in valid_actions}
+
+    def get_max(self, state: str, valid_actions: Optional[Sequence[int]] = None):
+        """get maximum q score for a state over actions"""
+        if valid_actions:
+            self.prune_qdict(state, valid_actions)
+        return max(self.qdict[state].values())
+
+    def get_arg_max(self, state, valid_actions: Optional[Sequence[int]] = None):
+        """
+        which action gives max q score. If identical max q scrores exist,
+           this'll choose somehing at random
+        """
+        mx = self.get_max(state, valid_actions)
+        # get all moves that return optimal outcome
+        all_arg_maxes = [k for k, v in self.qdict[state].items() if v == mx]
+        return random.choice(all_arg_maxes)
+
 
-        test_window = int(num_epsiodes / 20)
+class TicTacToeRunner:
+    def __init__(self, agent1: Agent, agent2: Agent):
+        self.env = tictactoe.TicTacToeEnv()
+        # player order
+        self.agent1 = agent1
+        self.agent2 = agent2
 
-        num_steps = np.zeros(num_epsiodes)
-        eps_vals = np.zeros(num_epsiodes)
+        # interact with the same environment, but through their preferred representation
+        self.agent1_env = self.agent1.wrapper(self.env)
+        # player 2 wrapper multiplies reward by -1
+        self.agent2_env = self.agent2.wrapper(self.env)
+
+    def train(self, num_epsiodes):
+        """"""
```

Comment:
This train method is key.  I've had a bit of a look up and down it, but I'd be keen maybe to spend 5 or 10 mins talking through it.